cmd/shell.go

```diff
@@ -48,68 +48,75 @@ func (m mysqlLogger) Print(v ...interface{}) {
 	log.Error(args...)
 	os.Exit(1)
 }
 func (log grpcLogger) Fatalf(format string, args ...interface{}) {
-	log.Error(args...)
+	log.Errorf(format, args...)
 	os.Exit(1)
 }
 func (log grpcLogger) Fatalln(args ...interface{}) {
-	log.Error(args...)
+	log.Errorln(args...)
 	os.Exit(1)
 }
 
-func (log grpcLogger) error(output string) {
-	if output == `ccResolverWrapper: error parsing service config: no JSON service config provided` {
-		return
-	}
-	if strings.Contains(output, `Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"`) {
-		return
-	}
-	log.Logger.AuditErr(output)
-}
-
+// Treat all gRPC error logs as potential audit events.
 func (log grpcLogger) Error(args ...interface{}) {
-	log.error(fmt.Sprint(args...))
+	log.Logger.AuditErr(fmt.Sprint(args...))
 }
 func (log grpcLogger) Errorf(format string, args ...interface{}) {
-	log.error(fmt.Sprintf(format, args...))
+	log.Logger.AuditErrf(format, args...)
 }
 func (log grpcLogger) Errorln(args ...interface{}) {
-	log.error(fmt.Sprint(args...))
+	log.Logger.AuditErr(fmt.Sprintln(args...))
 }
 
+// Pass through most Warnings, but filter out a few noisy ones.
 func (log grpcLogger) Warning(args ...interface{}) {
-	log.Error(args...)
+	log.Logger.Warning(fmt.Sprint(args...))
 }
 func (log grpcLogger) Warningf(format string, args ...interface{}) {
-	log.Errorf(format, args...)
+	log.Logger.Warningf(format, args...)
 }
 func (log grpcLogger) Warningln(args ...interface{}) {
-	log.Errorln(args...)
+	msg := fmt.Sprintln(args...)
+	// See https://github.com/letsencrypt/boulder/issues/4628
+	if strings.Contains(msg, `ccResolverWrapper: error parsing service config: no JSON service config provided`) {
+		return
+	}
+	// See https://github.com/letsencrypt/boulder/issues/4379
+	if strings.Contains(msg, `Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"`) {
+		return
+	}
+	// Since we've already formatted the message, just pass through to .Warning()
+	log.Logger.Warning(msg)
 }
 
 // Don't log any INFO-level gRPC stuff. In practice this is all noise, like
 // failed TXT lookups for service discovery (we only use A records).
-func (log grpcLogger) Info(args ...interface{}) {
-}
-func (log grpcLogger) Infof(format string, args ...interface{}) {
-}
-func (log grpcLogger) Infoln(args ...interface{}) {
+func (log grpcLogger) Info(args ...interface{})                 {}
+func (log grpcLogger) Infof(format string, args ...interface{}) {}
+func (log grpcLogger) Infoln(args ...interface{})               {}
+
+// V returns true if the verbosity level l is less than the verbosity we want to
+// log at.
+func (log grpcLogger) V(l int) bool {
+	// We always return false. This causes gRPC to not log some things which are
+	// only logged conditionally if the logLevel is set below a certain value.
+	// TODO: Use the wrapped log.Logger.stdoutLevel and log.Logger.syslogLevel
+	// to determine a correct return value here.
+	return false
```

Comment:
This is worth noticing. I'm *pretty sure* that our implementation of `V()` here has been incorrect. The semantics of how this method is used appears to be "return true if the set log level is such that we would print a statement at this given log level". Since log levels are all positive integers (with 0 representing error logs and higher integers representing less-critical logs), `l < 0` will always be false. This means that when the gRPC package asks:
```
	if logger.V(2) {
		logger.Info("base.baseBalancer: got new ClientConn state: ", s)
	}
```
we always reply "no, the set log level is not such that a log line at level 2 would get logged" and therefore it skips trying to log that line.

Here I've simplified our `l < 0` to `false`, to make it explicit that this is our current behavior, but I've also added a TODO to try to fix this so we respond correctly.

Reply:
I suspect we want to keep it as always false. We collect info-level logs from Boulder components, but the info-level logs you get out of gRPC if you allow them are verry verrbose. You can try it locally and see!

Reply:
Yeah, but gRPC doesn't *always* pass `2` to this function, so it's not just about whether or not we log info-level logs.

Reply:
As an SRE, would we have a way to make these logs show up when we're debugging a particularly annoying problem?

Reply:
As of right now, no, but also they've never shown up before. That's a good idea to include in the future, though.